packages/babel-generator/src/printer.ts

```diff
@@ -353,47 +367,45 @@ class Printer {
 
     this._maybePrintInnerComments(str, occurrenceCount);
 
-    const lastChar = this.getLastChar();
-    const strFirst = str.charCodeAt(0);
-    if (
-      (lastChar === charCodes.exclamationMark &&
-        // space is mandatory to avoid outputting <!--
-        // http://javascript.spec.whatwg.org/#comment-syntax
-        (str === "--" ||
-          // Needs spaces to avoid changing a! == 0 to a!== 0
-          strFirst === charCodes.equalsTo)) ||
-      // Need spaces for operators of the same kind to avoid: `a+++b`
-      (strFirst === charCodes.plusSign && lastChar === charCodes.plusSign) ||
-      (strFirst === charCodes.dash && lastChar === charCodes.dash) ||
-      // Needs spaces to avoid changing '34' to '34.', which would still be a valid number.
-      (strFirst === charCodes.dot && this._endsWithInteger)
-    ) {
-      this._space();
-    }
-
     this._maybeAddAuxComment();
-    this._append(str, maybeNewline, occurrenceCount);
+    this._append(str, maybeNewline, occurrenceCount, () => {
+      const lastChar = this.getLastChar();
+      const strFirst = str.charCodeAt(0);
+
+      return (
+        (lastChar === charCodes.exclamationMark &&
+          // space is mandatory to avoid outputting <!--
+          // http://javascript.spec.whatwg.org/#comment-syntax
+          (str === "--" ||
+            // Needs spaces to avoid changing a! == 0 to a!== 0
+            strFirst === charCodes.equalsTo)) ||
+        // Need spaces for operators of the same kind to avoid: `a+++b`
+        (strFirst === charCodes.plusSign && lastChar === charCodes.plusSign) ||
+        (strFirst === charCodes.dash && lastChar === charCodes.dash) ||
+        // Needs spaces to avoid changing '34' to '34.', which would still be a valid number.
+        (strFirst === charCodes.dot && this._endsWithInteger)
+      );
+    });
     this._noLineTerminator = false;
   }
 
   tokenChar(char: number): void {
     this.tokenContext = 0;
 
-    this._maybePrintInnerComments(String.fromCharCode(char));
-
-    const lastChar = this.getLastChar();
-    if (
-      // Need spaces for operators of the same kind to avoid: `a+++b`
-      (char === charCodes.plusSign && lastChar === charCodes.plusSign) ||
-      (char === charCodes.dash && lastChar === charCodes.dash) ||
-      // Needs spaces to avoid changing '34' to '34.', which would still be a valid number.
-      (char === charCodes.dot && this._endsWithInteger)
-    ) {
-      this._space();
-    }
+    const str = String.fromCharCode(char);
+    this._maybePrintInnerComments(str);
 
     this._maybeAddAuxComment();
-    this._appendChar(char);
+    this._appendChar(char, () => {
```

Comment:
This change has also the effect that we don't inject the space anymore if it's made unnecessary by the aux comment (e.g. `34/* asux comment*/.foo`).